### PR TITLE
Improve swing foot plan

### PR DESCRIPTION
**Pull request by yanhaoy**
Remove online update for swing foot trajectory
Minimum circle search for foothold
Bug fix

### CODE REVIEW

File: local_planner/include/local_planner/local_footstep_planner.h [hunk 1]
```diff
@@ -240,6 +240,129 @@ class LocalFootstepPlanner {
...
+        std::min(ground_clearance_ - toe_radius +
+                     std::max(foot_position_prev.z(), foot_position_next.z()),
+                 hip_height - hip_clearance_);
+
+    return swing_apex;
+  }
+
+  // Compute future states by integrating linear states (hold orientation
+  // states)
+  inline Eigen::VectorXd computeFutureState(double step,
```
Comment (ardalantj) [inline local_planner/include/local_planner/local_footstep_planner.h:265]: computeFutureBodyPlan?

File: local_planner/local_planner.yaml [hunk 2]
```diff
@@ -27,8 +27,8 @@ local_footstep_planner:
   foothold_search_radius: 0.2
   foothold_obj_threshold: 0.6 # Note: For now must match /grid_map_filters/traversability_mask_lower_threshold/lower_threshold
   obj_fun_layer: traversability
-  # duty_cycles: [0.3, 0.3, 0.3, 0.3]
-  duty_cycles: [0.5, 0.5, 0.5, 0.5]
+  duty_cycles: [0.3, 0.3, 0.3, 0.3]
+  # duty_cycles: [0.5, 0.5, 0.5, 0.5]
```
Comment (ardalantj) [inline local_planner/local_planner.yaml:31]: remove?

File: local_planner/src/local_footstep_planner.cpp [hunk 3]
```diff
@@ -280,6 +285,29 @@ void LocalFootstepPlanner::computeFootPlan(
         getFootData(foot_positions, i_touchdown, j);
     Eigen::Vector3d foot_velocity_next = Eigen::Vector3d::Zero();
 
+    // Compute the midair index
+    int midair = std::round(i_liftoff + swing_duration / 2);
```
Comment (ardalantj) [inline local_planner/src/local_footstep_planner.cpp:289]: mid_air?

File: local_planner/src/local_planner.cpp [hunk 4]
```diff
@@ -428,9 +417,9 @@ void LocalPlanner::getStateAndTwistInput() {
   ref_body_plan_(0, 11) = cmd_vel_[5];
 
   // Only adaptive pitch
-  ref_body_plan_(0, 4) = local_footstep_planner_->getTerrainSlope(
-      current_state_(0), current_state_(1), current_state_(6),
-      current_state_(7));
+  // ref_body_plan_(0, 4) = local_footstep_planner_->getTerrainSlope(
```
Comment (ardalantj) [inline local_planner/src/local_planner.cpp:420]: remove?

File: local_planner/src/local_planner.cpp [hunk 5]
```diff
@@ -461,9 +450,9 @@ void LocalPlanner::getStateAndTwistInput() {
     ref_body_plan_(i, 2) = z_des_ + ref_ground_height_(i);
 
     // Only adaptive pitch
-    ref_body_plan_(i, 4) = local_footstep_planner_->getTerrainSlope(
-        ref_body_plan_(i, 0), ref_body_plan_(i, 1), ref_body_plan_(i, 6),
-        ref_body_plan_(i, 7));
+    // ref_body_plan_(i, 4) = local_footstep_planner_->getTerrainSlope(
```
Comment (ardalantj) [inline local_planner/src/local_planner.cpp:453]: remove?

File: local_planner/include/local_planner/local_footstep_planner.h [hunk 6]
```diff
@@ -240,6 +240,129 @@ class LocalFootstepPlanner {
     terrain_grid_ = grid_map;
   }
 
+  // Compute swing apex by ground and hip height
+  inline double computeSwingApex(int leg_idx, const Eigen::VectorXd &body_plan,
```
Comment (ardalantj) [inline local_planner/include/local_planner/local_footstep_planner.h:244]: Any reason these functions should be in the header file?

File: local_planner/include/local_planner/local_footstep_planner.h (hunk 1 above)
Comment (yanhaoy) [reply local_planner/include/local_planner/local_footstep_planner.h:265]: Sure

File: local_planner/src/local_planner.cpp (hunk 4 above)
Comment (yanhaoy) [reply local_planner/src/local_planner.cpp:420]: I want to keep them. There are two ways to align with terrain, here is another way.

File: local_planner/src/local_planner.cpp (hunk 5 above)
Comment (yanhaoy) [reply local_planner/src/local_planner.cpp:453]: I want to keep them. There are two ways to align with terrain, here is another way.

File: local_planner/src/local_planner.cpp (hunk 4 above)
Comment (ardalantj) [reply local_planner/src/local_planner.cpp:420]: Ok, maybe just add a comment to clarify that

File: local_planner/include/local_planner/local_footstep_planner.h (hunk 6 above)
Comment (ardalantj) [reply local_planner/include/local_planner/local_footstep_planner.h:244]: Normally inlines are used for functions with small definitions to save time accessing the implementation of the function when it is called. I think these can be moved to cpp.

**Comment by ardalantj** [comment]
Is there any additional test needed to cover this specific feature?

File: local_planner/include/local_planner/local_footstep_planner.h (hunk 6 above)
Comment (jcnorby) [reply local_planner/include/local_planner/local_footstep_planner.h:244]: Yeah ideally inlines should be only a few lines or smaller. We may have gotten a bit aggressive in other parts of the code...

**Comment by yanhaoy** [comment]
> Is there any additional test needed to cover this specific feature?

I think it's fine, this PR is mainly to fix bugs. The only new feature might be the minimal circle, but the foothold test should already have that covered.

**Comment by ardalantj** [comment]
Looks good to me. Feel free to merge once pipeline is passing.

File: local_planner/src/local_planner.cpp [hunk 7]
```diff
@@ -385,25 +393,6 @@ void LocalPlanner::getStateAndTwistInput() {
...
-    stand_pose_ << current_state_[0], current_state_[1], current_state_[5];
-  } else {
-    // If it's standing, try to stablized the waggling
-    Eigen::Vector3d current_stand_pose;
-    current_stand_pose << current_state_[0], current_state_[1],
-        current_state_[5];
-    stand_pose_ = stand_pose_ * (1 - 1 / update_rate_) +
-                  current_stand_pose * 1 / update_rate_;
-  }
-
```
Comment (jcnorby) [inline local_planner/src/local_planner.cpp:406]: Can you revert these changes? The next PR will change these sections to enable a standing mode, so changes to this will cause conflicts.

**Review by jcnorby — CHANGES_REQUESTED** [review]

**Review by ardalantj — APPROVED** [review]

File: local_planner/src/local_footstep_planner.cpp [hunk 8]
```diff
@@ -174,47 +171,67 @@ void LocalFootstepPlanner::computeFootPlan(
...
-
-        // Compute body pose at midstance from either the trajectory or Raibert
-        // heuristic directly
-        if (midstance < horizon_length_) {
-          body_pos_midstance = body_plan.block<1, 3>(midstance, 0);
-          body_rpy_midstance = body_plan.block<1, 3>(midstance, 3);
-          // grf_midstance = grf_plan.block<1, 3>(midstance, 3 * j);
+        // Compute the number of timesteps corresponding to the end of stance
+        // phase
+        int end_of_stance = i + period_ * duty_cycles_[j];
```
Comment (jcnorby) [inline local_planner/src/local_footstep_planner.cpp:176]: We probably want this to use the contact schedule rather than the period/duty cycles, similar to the already existing getNextContactState.

File: local_planner/src/local_footstep_planner.cpp [hunk 9]
```diff
@@ -174,47 +171,67 @@ void LocalFootstepPlanner::computeFootPlan(
...
+          Eigen::Vector2d p;
+          p << hip_position_midstance.x(), hip_position_midstance.y();
+
+          bool duplicate = false;
+          for (size_t l = 0; l < P.size(); l++) {
+            if ((P.at(l) - p).norm() < 1e-3) {
+              duplicate = true;
+              break;
+            }
+          }
```
Comment (jcnorby) [inline local_planner/src/local_footstep_planner.cpp:208]: What is the reason for doing this?

File: local_planner/src/local_footstep_planner.cpp [hunk 10]
```diff
@@ -174,47 +171,67 @@ void LocalFootstepPlanner::computeFootPlan(
...
+        // Ref: Highly Dynamic Quadruped Locomotion via Whole-Body Impulse
+        // Control and Model Predictive Control (Centrifugal force and capture
+        // point)
+        centrifugal = body_height_touchdown / 9.81 *
                       body_vel_touchdown.cross(ref_body_ang_vel_touchdown);
-        vel_tracking =
-            (hip_height / 9.81) * (body_vel_touchdown - ref_body_vel_touchdown);
+        // Ref: MIT Cheetah 3: Design and Control of a Robust, Dynamic Quadruped
+        // Robot (Capture point)
+        vel_tracking = std::sqrt(body_height_touchdown / 9.81) *
```
Comment (jcnorby) [inline local_planner/src/local_footstep_planner.cpp:232]: Didn't we decide that this didn't need a sqrt?

File: local_planner/src/local_footstep_planner.cpp [hunk 11]
```diff
@@ -589,3 +577,92 @@ Eigen::Vector3d LocalFootstepPlanner::getNearestValidFoothold(
...
+  Eigen::Vector2d p = P.back();
+  P.pop_back();
+  Eigen::Vector3d D = welzlMinimumCircle(P, R);
+  if ((p - D.segment(0, 2)).norm() < D.z()) {
+    return D;
+  }
+
+  R.push_back(p);
+  return welzlMinimumCircle(P, R);
+}
```
Comment (jcnorby) [inline local_planner/src/local_footstep_planner.cpp:668]: Have we benchmarked how long computeFootPlan takes with these new additions?

File: local_planner/src/local_footstep_planner.cpp (hunk 9 above)
Comment (yanhaoy) [reply local_planner/src/local_footstep_planner.cpp:208]: The minimal circle algorithm cannot accept repeated inputs.

File: local_planner/src/local_footstep_planner.cpp (hunk 10 above)
Comment (yanhaoy) [reply local_planner/src/local_footstep_planner.cpp:232]: I think that's for the centrifugal term.

File: local_planner/src/local_footstep_planner.cpp (hunk 11 above)
Comment (yanhaoy) [reply local_planner/src/local_footstep_planner.cpp:668]: I can benchmark it. But this algorithm is recursive and the time complexity is O(N). Since we entered a stance plan of about 6 elements, it basically calculates the circumcircle of 3 points and computes the distance of the remaining points.

Comment (yanhaoy) [reply local_planner/src/local_footstep_planner.cpp:668]: The circumcircle looks complicated because I copied it directly from MATLAB. I suspect it might be more simplified if I generate matlabfunction.

File: local_planner/include/local_planner/local_footstep_planner.h [hunk 12]
```diff
@@ -240,6 +240,40 @@ class LocalFootstepPlanner {
...
```
Comment (ardalantj) [inline local_planner/include/local_planner/local_footstep_planner.h:244]: This should go to cpp?

File: local_planner/src/local_footstep_planner.cpp (hunk 11 above)
Comment (yanhaoy) [reply local_planner/src/local_footstep_planner.cpp:668]: It takes between 0.05 and 0.1 ms on my laptop.

**Review by jcnorby — APPROVED** [review]
Looks good - it was out of sync with the latest PR which added the stand mode and removed twist body planner. I pulled in those changes, checked the simulation, and pushed, @yanhaoy go ahead and merge if everything looks good
